hnn_core/lfp.py

```diff
@@ -0,0 +1,407 @@
+        else:
+            return None  # simulation not yet run
+
+    @property
+    def times(self):
+        """The sampling time points."""
+        if self._lfp_t.size() > 0:
+            # NB _lfp_t is one sample longer than _lfp_v
+            return self._lfp_t.to_python()[:self.n_samples]
+        else:
+            return None  # simulation not yet run
+
+    def build(self):
+        """Create the Neuron objects needed to record LFPs
+
+        """
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self.imem_ptrvec = h.PtrVector(segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        self.imem_vec_len = int(self.imem_ptrvec.size())
+        self.imem_vec = h.Vector(self.imem_vec_len)
+
+        ptr_count = 0
+        for sec in secs_on_rank:
+            for seg in sec:  # section end points (0, 1) not included
+                # set Nth pointer to the net membrane current at this segment
+                self.imem_ptrvec.pset(
+                    ptr_count, sec(seg.x)._ref_i_membrane_)
+                ptr_count += 1
+        if ptr_count != self.imem_vec_len:
+            raise RuntimeError(f'Expected {self.imem_vec_len} imem pointers, '
+                               f'got {ptr_count}.')
+
+        # transfer resistances for each segment (keep in Neuron Matrix object)
+        self.r_transfer = h.Matrix(self.n_contacts, self.imem_vec_len)
+
+        for row, pos in enumerate(self.positions):
+            transfer_resistance = list()
+            for sec, n_segs in zip(secs_on_rank, segment_counts):
+                this_xfer_r = _transfer_resistance(sec, pos,
+                                                   sigma=self.sigma,
+                                                   method=self.method)
+                # the n_segs of this section get assigned the same value (e.g.
+                # in PSA, the distance is calculated for the section mid point
+                # only)
+                transfer_resistance.extend([this_xfer_r] * n_segs)
+
+            self.r_transfer.setrow(row, h.Vector(transfer_resistance))
+
+        # record time for each array
+        self._lfp_t = h.Vector().record(h._ref_t)
+
+        # contributions of all segments on this rank to total calculated
+        # potential at electrode (_PC.allreduce called in _simulate_dipole)
+        self.reset()
+
+    def reset(self):
+        self._lfp_v = h.Vector()
+        self.imem_vec = h.Vector(self.imem_vec_len)
+
+    def calc_potential_callback(self):
+        # keep all data in Neuron objects for efficiency
+
+        # 'gather' the values of seg.i_membrane_ into self.imem_vec
```

Comment:
Just a comment. I think some of these terms should go into a glossary ... "gather", "gid", "reduce", "rank" etc.

Reply:
Not sure I follow. Glossary? In the docstring?

Reply:
Not in the docstring. Once we have a page for glossary ... was a tangential comment :)